### PR TITLE
Configurable word size for test signature file output

### PR DESCRIPTION
**Pull request by bilalsakhawat-10xe**
Fix for the issue #111 
Now by default Sail with dump 32-bit (word) length signatures in the signature file.
To make it configurable `signature-granularity=<num-bytes-per-line>` cli is added, just like we have in spike.

### CODE REVIEW

**Comment by abukharmeh** [comment]
This raises the question if we want to maintain consistency between c and ocaml emulators @martinberger and what is the expected usage of each of them !

**Comment by martinberger** [comment]
> This raises the question weather we want to maintain consistency between c and ocaml emulators @martinberger and what is the expected usage of each of them !

@abukharmeh Good question. As far as I understand Sail's history, the  purpose of the OCaml emulator was to bootstrap the C emulator, since it's much easier to get OCaml correct than C. The OCaml emulator is quite a lot slower than the C emulator. (@PeterSewell @Alasdair please correct me if I'm wrong)

**Comment by allenjbaum** [comment]
In https://github.com/riscv/sail-riscv/issues/111, jrtc27 commented on Oct 4
Oh ok that thing (which also exists for the OCaml simulator, I might add). Making that configurable isn't hard.

Which implies to me that it is already in the Ocaml simulator - so that shouldn't be an issue.

**Comment by abukharmeh** [comment]
Yes its OCaml emulators, but is not configurable (as far as I can see). If we want to maintain consistency between emulators, then this function should be changed.

https://github.com/riscv/sail-riscv/blob/5549f159b7b7f1fd0e27b37202eee0d103661e9e/ocaml_emulator/riscv_ocaml_sim.ml#L109-L119

**Comment by allenjbaum** [comment]
OK, that's odd since (I interpreted) jrtc27's comment
 " Oh ok that thing (which also exists for the OCaml simulator, I might
add). Making that configurable isn't hard."
to mean that it was already in OCaml, but I can see how it might mean that
the same implementation that is the C-model is also in the OCaml model, and
therefore should also be fixed.

**Comment by jrtc27** [comment]
Ibrahim's interpretation is what I was trying to convey; I specifically called out the OCaml version as also having the test signature dumping feature, since https://github.com/riscv/sail-riscv/issues/111#issuecomment-933673462 only referenced the C version, and wanted to make sure we kept feature parity.

**Comment by allenjbaum** [comment]
Thanks; I naturally interpreted exactly the opposite. It's one of my
superpowers.
And why we need more formal definitions....

**Comment by allenjbaum** [comment]
So, we have a PR for the CSIM support; the question is whether we should
fix both C-sim and OCaml in the same PR or separate ones.
Any opinions?

**Comment by bilalsakhawat-10xe** [comment]
@abukharmeh @jrtc27  I made similar changes to ocaml emulator/simulator. It's my first time writing ocaml syntax so suggestions are welcome.

File: ocaml_emulator/riscv_ocaml_sim.ml [hunk 1]
```diff
@@ -109,13 +116,15 @@ let check_elf () =
...
   while !i < Bytes.length bytes do
-    let s = Printf.sprintf "%02x%02x%02x%02x\n"
-	(int_of_char (Bytes.get bytes (!i+3)))
-	(int_of_char (Bytes.get bytes (!i+2)))
-	(int_of_char (Bytes.get bytes (!i+1)))
-	(int_of_char (Bytes.get bytes (!i+0))) in
-    output_string fl s;
-    i := !i + 4
+    let j = ref 0 in
+    while !j < !signature_granularity_ref do
```
Comment (martinberger) [inline ocaml_emulator/riscv_ocaml_sim.ml:120]: Why not use a `for`-loop here, then you can avoid hand-incrementing the counter `j`

Comment (martinberger) [reply ocaml_emulator/riscv_ocaml_sim.ml:120]: Something like

    for j = 0 to !signature_granularity_ref - 1 do
        ...
    done

Comment (martinberger) [reply ocaml_emulator/riscv_ocaml_sim.ml:120]: I don't think that OCaml has a `for` loop by increments other than 1 and -1, so the other loop can't be a `for`-loop alas

File: c_emulator/riscv_sim.c [hunk 2]
```diff
@@ -68,6 +68,7 @@ size_t spike_dtb_len = 0;
 char *sig_file = NULL;
 uint64_t mem_sig_start = 0;
 uint64_t mem_sig_end = 0;
+uint64_t signature_granularity = 4;
```
Comment (abukharmeh) [inline c_emulator/riscv_sim.c:71]: Why the size of this is 64 bit, and then you are using int in the for loop ?

Comment (abukharmeh) [reply c_emulator/riscv_sim.c:71]: Apart from that, I dont have anything else to comment on, and this looks and works fine to me.

Comment (abukharmeh) [reply c_emulator/riscv_sim.c:71]: Oh actually, it might be helpfull to assert that signature_granularity is divisable by 4 !

Comment (bilalsakhawat-10xe) [reply c_emulator/riscv_sim.c:71]: sure I'll change the signature_granularity size to int.
About "Δ mem_sig is divisable by signature_granularity" can you please elaborate it a little? like when Δ mem_sig is 
 not divisable by signature_granularity, is the signature granularity is invalid or Δ mem_sig.

File: ocaml_emulator/riscv_ocaml_sim.ml (hunk 1 above)
Comment (bilalsakhawat-10xe) [reply ocaml_emulator/riscv_ocaml_sim.ml:120]: So, I can only change the inner while loop to a for loop, right? I'll do it in the next commit.

File: c_emulator/riscv_sim.c (hunk 2 above)
Comment (abukharmeh) [reply c_emulator/riscv_sim.c:71]: Maybe this is unnecessary, do we know for sure that the addresses are word-aligned or is there any alignment restriction on them? In the last iteration of the loop, if addr < mem_sig_end, then we will increment by signature_granularity, which could make our address outside the signature region?

**Comment by allenjbaum** [comment]
If the signature isn't aligned, then it's not a Sail bug, it's a test bug.
Also the SIGUPD macros are supposed to keep the signature pointer XLEN
aligned

File: c_emulator/riscv_sim.c [hunk 3]
```diff
@@ -578,10 +585,10 @@ void write_signature(const char *file)
     fprintf(stderr, "Cannot open file '%s': %s\n", file, strerror(errno));
     return;
   }
-  /* write out words in signature area */
-  for (uint64_t addr = mem_sig_start; addr < mem_sig_end; addr += 4) {
+  /* write out words depending on signature granularity in signature area */
+  for (uint64_t addr = mem_sig_start; addr < mem_sig_end; addr += signature_granularity) {
     /* most-significant byte first */
-    for (int i = 3; i >= 0; i--) {
+    for (int i = signature_granularity - 1 ; i >= 0; i--) {
```
Comment (jrtc27) [inline c_emulator/riscv_sim.c:591]:
```suggestion
    for (int i = signature_granularity - 1; i >= 0; i--) {
```

File: ocaml_emulator/riscv_ocaml_sim.ml [hunk 4]
```diff
@@ -109,13 +116,13 @@ let check_elf () =
...
-	(int_of_char (Bytes.get bytes (!i+0))) in
-    output_string fl s;
-    i := !i + 4
+    for j = 0 to !signature_granularity_ref - 1 do
+      let s = Printf.sprintf "%02x"
+          (int_of_char (Bytes.get bytes (!i + (!signature_granularity_ref - 1 - j)))) in
+      output_string fl s;
+    done;
+    i := !i + !signature_granularity_ref;
+    output_string fl "\n";
```
Comment (jrtc27) [inline ocaml_emulator/riscv_ocaml_sim.ml:125]: Probably nicer to keep the original order, i.e. swap these two lines back?

File: ocaml_emulator/riscv_ocaml_sim.ml [hunk 5]
```diff
@@ -61,6 +65,9 @@ let options = Arg.align ([("-dump-dts",
                           ("-test-signature",
                            Arg.String set_signature_file,
                            " file for signature output (requires ELF signature symbols)");
+                          ("-signature-granularity",
+                           Arg.Int set_signature_granularity,
+                           " test signature granularity (in Bytes)");
```
Comment (jrtc27) [inline ocaml_emulator/riscv_ocaml_sim.ml:70]:
```suggestion
                           " test signature granularity (in bytes)");
```
unless Bytes specifically is used elsewhere that I can't see (I can see MB but that's different)

File: ocaml_emulator/riscv_ocaml_sim.ml [hunk 6]
```diff
@@ -109,13 +116,13 @@ let check_elf () =
...
-    let s = Printf.sprintf "%02x%02x%02x%02x\n"
-	(int_of_char (Bytes.get bytes (!i+3)))
-	(int_of_char (Bytes.get bytes (!i+2)))
-	(int_of_char (Bytes.get bytes (!i+1)))
-	(int_of_char (Bytes.get bytes (!i+0))) in
-    output_string fl s;
-    i := !i + 4
+    for j = 0 to !signature_granularity_ref - 1 do
+      let s = Printf.sprintf "%02x"
+          (int_of_char (Bytes.get bytes (!i + (!signature_granularity_ref - 1 - j)))) in
```
Comment (jrtc27) [inline ocaml_emulator/riscv_ocaml_sim.ml:121]: You could use `1 to !signature_granularity_ref` for the loop and drop the - 1 here, or just `for j = !signature_granularity_ref - 1 downto 0` like the C version and not have to do the inversion at all

**Comment by bilalsakhawat-10xe** [comment]
@jrtc27 I have made the suggested changes.

**Review by jrtc27 — APPROVED** [review]
Thanks

**Review by martinberger — APPROVED** [review]
Looks fine.

**Comment by martinberger** [comment]
Merged!

Thanks @bilalsakhawat-10xe @abukharmeh @jrtc27

**Pull request merged**